src/Evolution/Systems/ScalarAdvection/Subcell/TimeDerivative.hpp

```diff
@@ -0,0 +1,246 @@
+#include "Evolution/Systems/ScalarAdvection/Fluxes.hpp"
+#include "Evolution/Systems/ScalarAdvection/Subcell/ComputeFluxes.hpp"
+#include "Evolution/Systems/ScalarAdvection/System.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "Utilities/ErrorHandling/Assert.hpp"
+#include "Utilities/FakeVirtual.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/MakeArray.hpp"
+
+namespace ScalarAdvection::subcell {
+/*!
+ * \brief Compute the time derivative on the subcell grid using FD
+ * reconstruction.
+ *
+ * The code makes the following unchecked assumptions:
+ * - Assumes Cartesian coordinates with a diagonal Jacobian matrix
+ * from the logical to the inertial frame
+ * - Assumes the mesh is not moving (grid and inertial frame are the same)
+ */
+template <size_t Dim>
+struct TimeDerivative {
+  template <typename DbTagsList>
+  static void apply(
+      const gsl::not_null<db::DataBox<DbTagsList>*> box,
+      const InverseJacobian<DataVector, Dim, Frame::ElementLogical,
+                            Frame::Grid>&
+          cell_centered_logical_to_grid_inv_jacobian,
+      const Scalar<DataVector>& /*cell_centered_det_inv_jacobian*/) {
+    // subcell is currently not supported for external boundary elements
+    const Element<Dim>& element = db::get<domain::Tags::Element<Dim>>(*box);
+    ASSERT(element.external_boundaries().size() == 0,
+           "Can't have external boundaries right now with subcell. ElementID "
+               << element.id());
+
+    using evolved_vars_tags = typename System<Dim>::variables_tag::tags_list;
+    using fluxes_tags = typename Fluxes<Dim>::return_tags;
+
+    // The copy of Mesh is intentional to avoid a GCC-7 internal compiler error.
+    const Mesh<Dim> subcell_mesh =
+        db::get<evolution::dg::subcell::Tags::Mesh<Dim>>(*box);
+    ASSERT(
+        subcell_mesh == Mesh<Dim>(subcell_mesh.extents(0),
+                                  subcell_mesh.basis(0),
+                                  subcell_mesh.quadrature(0)),
+        "The subcell/FD mesh must be isotropic for the FD time derivative but "
+        "got "
+            << subcell_mesh);
+
+    const size_t num_reconstructed_pts =
+        (subcell_mesh.extents(0) + 1) *
+        subcell_mesh.extents().slice_away(0).product();
+
+    const ScalarAdvection::fd::Reconstructor<Dim>& recons =
+        db::get<ScalarAdvection::fd::Tags::Reconstructor<Dim>>(*box);
+
+    const auto& boundary_correction =
+        db::get<evolution::Tags::BoundaryCorrection<System<Dim>>>(*box);
+    using derived_boundary_corrections =
+        typename std::decay_t<decltype(boundary_correction)>::creatable_classes;
+
+    // Variables to store the boundary correction terms on FD subinterfaces
+    std::array<Variables<evolved_vars_tags>, Dim> boundary_corrections{};
+
+    // package the data and compute the boundary correction
+    tmpl::for_each<derived_boundary_corrections>([&](auto
```

Comment:
gcc7 has a compiler bug that can be avoided by specifying the variables to be captured (instead of all of them as is done here)